Author: Sami Mwanje, mm223kk@student.lnu.se

The code and structure is based from the pages:

https://customers.pyimagesearch.com/lesson-sample-k-nearest-neighbor-classification/

https://medium.com/analytics-vidhya/a-beginners-guide-to-knn-and-mnist-handwritten-digits-recognition-using-knn-from-scratch-df6fb982748a


In [1]:
import numpy as np
import  pandas as pd
from MinstHelp import knn


Import all the needed labaries and external files.

Numpy for the matrixes.

Pandas to faster read the dataset files csv files.

MinistHelp which is used for the regression.

In [2]:
mnistTrain = np.asarray(pd.read_csv('mnistTrain.csv', delimiter=',')) # Read the minist data file.
mnistTest = np.asarray(pd.read_csv('mnistTest.csv', delimiter=',')) # Read the minist data file.
kValues = [1,3,5,7,9,11]  # Decision boundary with corresponding training error for k

trainingSet_X = np.array(mnistTrain[:10000,1:])
trainingSet_Y = np.array(mnistTrain[:10000,:1])
print('Training set X: %s,\nTraining set Y: %s' % (trainingSet_X.shape, trainingSet_Y.shape )) # print the dimentions of the train data

#Divide the dataset into a training set of size.
testSet_X = np.array(mnistTest[:1000,1:])
testSet_Y = np.array(mnistTest[:1000,:1])
print('\nTest set X: %s,\nTest set Y: %s' % (testSet_X.shape, testSet_Y.shape )) # print the dimentions of the train data

Training set X: (10000, 784),
Training set Y: (10000, 1)

Test set X: (1000, 784),
Test set Y: (1000, 1)


The training set and test set is created from the data set using the pandas libarary.
Kvalues 1,3,5,7,9 and 11 are created and are going to be using for classification.


The data is splited up into trainig set and test set.

Each set is split up into X-part, and Y-part.

Y-part containts the labels, and the X-part contains the features for the MNIST data set.


In [3]:
print('Training results:')         # Training set results.
for k in kValues:
    accuracy = len(trainingSet_X)                                               # Accuracy variable as large as the the data set rows.
    for x in range(len(trainingSet_X)):
        result = knn(trainingSet_X, trainingSet_X[x], trainingSet_Y, trainingSet_Y, k, x)      # Calculate ecdulian distance and return the knn result.
        if result == 0: 
            accuracy -= 1                                                       # Decrease accuracy variable on failed knn.
    
    procent =  np.round( (100*(accuracy/len(trainingSet_X))), 2)                # Calculate the success rate.
    print('Accuracy Training K=' + str(k) + ': ' + str(procent) + ' %')

Training results:


KeyboardInterrupt: 

The MinstHelp external functions are used here to calculate the ecludian distance,
and return if the knn result.

Data from the training set is used to calculate the nearest neighbours of other datain the training set.
If the the knn has the correct majority accuracy will remain otherwise the accuracy will decrease.

In [ ]:
print('\nTest results:')        # Test set results.
for k in kValues:
    accuracy = len(testSet_X)                                      # Accuracy variable as large as the the data set rows.       
    for x in range(len(testSet_X)):
        result = knn(trainingSet_X, testSet_X[x], testSet_Y, trainingSet_Y, k, x)    # Calculate ecdulian distance and return the knn result.
        if result == 0:
            accuracy -= 1                                         # Decrease accuracy variable on failed knn.
    
    procent =  np.round( (100*(accuracy/len(testSet_X))), 2)     # Calculate the success rate.
    print('Accuracy Test K=' + str(k) + ': ' + str(procent) + ' %')


Test results:
Accuracy Test K=1: 92.0 %
Accuracy Test K=3: 91.6 %
Accuracy Test K=5: 90.8 %
Accuracy Test K=7: 90.9 %
Accuracy Test K=9: 90.5 %
Accuracy Test K=11: 89.5 %


Knn was used on both training and test set. The k values [1,3,7,9,11] were used.

Using K = 3, gave the best accuracy for both training and test set. Which is 97 vs 91 %.

Though K = has a 96.15 % training accuracy and a 91.6 % test accuracy, which is a lower variance.
Whith still a high accuracy.  

In the test set we could see a larger difference between k = 1 and k = 11. This was,
almost a 10 % difference. Going to a higher K than 3-5 is therefore not preferable for sensitive data.

Higher values of K gave a higher variance error, but lower bias error. 

<b>Training results:                           &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;        Test results:</b>  

<p> Accuracy Training K=1: 100.0 %             &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;                Accuracy Test K=1: 100.0 %</p>
<p>Accuracy Training K=3: 97.37 %              &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;                Accuracy Test K=3: 91.6 %</p>
<p>Accuracy Training K=5: 96.15 %              &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;                Accuracy Test K=5: 90.8 %</p>
<p>Accuracy Training K=7: 95.18 %              &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;                Accuracy Test K=7: 90.9 %</p>
<p>Accuracy Training K=9: 94.59 %              &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;                Accuracy Test K=9: 90.5 %</p>
<p>Accuracy Training K=11: 93.98 %             &emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&emsp;&nbsp;                Accuracy Test K=11: 89.5 %</p>